packages/service-library/src/servicelib/rabbitmq/_client.py

```diff
@@ -20,6 +21,63 @@
 
 _DEFAULT_PREFETCH_VALUE: Final[int] = 10
 _DEFAULT_RABBITMQ_EXECUTION_TIMEOUT_S: Final[int] = 5
+_HEADER_X_DEATH: Final[str] = "x-death"
+
+DEFAULT_UNEXPECTED_ERROR_RETRY_DELAY_S: Final[float] = 1
+DEFAULT_UNEXPECTED_ERROR_MAX_ATTEMPTS: Final[NonNegativeInt] = 15
+
+
+def _get_delayed_exchange_name(exchange_name: str) -> str:
+    return f"delayed_{exchange_name}"
+
+
+def _get_x_death_count(message: aio_pika.abc.AbstractIncomingMessage) -> int:
+    count: int = 0
+
+    x_death: list[dict[str, Any]] = message.headers.get(_HEADER_X_DEATH, [])
+    if x_death:
+        count = x_death[0]["count"]
+
+    return count
+
+
+async def _on_message(
+    message_handler: MessageHandler,
+    max_retries_upon_error: int,
+    message: aio_pika.abc.AbstractIncomingMessage,
+) -> None:
+    async with message.process(requeue=True, ignore_processed=True):
+        count = _get_x_death_count(message)
+        try:
+            with log_context(
+                _logger,
+                logging.DEBUG,
+                msg=f"Received message from {message.exchange=}, {message.routing_key=}",
+            ):
+                if not await message_handler(message.body):
+                    # NOTE: will put message to original queue(not the DLX)
+                    await message.nack()
+        except Exception:  # pylint: disable=broad-exception-caught
```

Comment:
Q: This exception could come from `message_handler` or `message.nack`. If it is the latter, isnt' there a risk that `await message.nack(requeue=False)` , used in the handler reraise  as well?

Reply:
Incorrect. This also got me a bit confused. I tried to make it clear with the notes and also wrote it in the description of the PR.
- `message.nack()` puts the message the original queue
- `message.nack(requeue=False)` puts the message in the dead letter exchange